### PR TITLE
Added the breadcrumb to the editmask

### PR DESCRIPTION
**Pull request by cboelter**
As discussed here https://github.com/contao-community-alliance/dc-general/issues/101 we can now add the breadcrumb to the editmask.

### CODE REVIEW

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php [hunk 1]
```diff
@@ -713,4 +716,63 @@ class_implements(
...
+        $arrReturn = $event->getElements();
+
+        if (!is_array($arrReturn) || count($arrReturn) == 0) {
+            return null;
+        }
+
+        $GLOBALS['TL_CSS'][] = 'system/modules/dc-general/html/css/generalBreadcrumb.css';
+
+        $objTemplate = $this->getTemplate('dcbe_general_breadcrumb');
+        $this->addToTemplate('elements', $arrReturn, $objTemplate);
```
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: Could you use `$template->set(/* .... */);` here?
The `addToTemplate()` calls are from a time where there was no setter method in the template yet and therefore anonymous properties on the template were set which we dislike.

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php [hunk 2]
```diff
@@ -713,4 +716,63 @@ class_implements(
...
+     * @SuppressWarnings(PHPMD.Superglobals)
+     * @SuppressWarnings(PHPMD.CamelCaseVariableName)
+     */
+    protected function breadcrumb()
+    {
+        $event = new GetBreadcrumbEvent($this->getEnvironment());
+
+        $this->getEnvironment()->getEventDispatcher()->dispatch(sprintf('%s', $event::NAME), $event);
+
+        $arrReturn = $event->getElements();
```
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: Please do not use type prefixes in variables in new code. We are changing the codebase to not use prefixes step by step. Therefore we should not introduce them with new code.

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php [hunk 3]
```diff
@@ -713,4 +716,63 @@ class_implements(
...
+    }
+
+    /**
+     * Create a new instance of ContaoBackendViewTemplate with the template file of the given name.
+     *
+     * @param string $strTemplate Name of the template to create.
+     *
+     * @return ContaoBackendViewTemplate
+     */
+    protected function getTemplate($strTemplate)
```
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: This method is only to be called once and there will not likely be any reusage of it. Please move the call to the code above.

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php [hunk 4]
```diff
@@ -713,4 +716,63 @@ class_implements(
...
+     *
+     * @param string    $name     Name of the value.
+     *
+     * @param mixed     $value    The value to add to the template.
+     *
+     * @param \Template $template The template to add the value to.
+     *
+     * @return BaseView
+     */
+    protected function addToTemplate($name, $value, $template)
```
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: This method will get obsolete when using `$template->set()`.

**Comment by discordier** [comment]
First off, good work. Thanks a lot.

Please change the mentioned stuff and I'd be happy to merge.

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php [hunk 5]
```diff
@@ -713,4 +716,63 @@ class_implements(
...
+     * @return string
+     *
+     * @SuppressWarnings(PHPMD.Superglobals)
+     * @SuppressWarnings(PHPMD.CamelCaseVariableName)
+     */
+    protected function breadcrumb()
+    {
+        $event = new GetBreadcrumbEvent($this->getEnvironment());
+
+        $this->getEnvironment()->getEventDispatcher()->dispatch(sprintf('%s', $event::NAME), $event);
```
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: The `sprintf` is not needed here? Why did you use it?

**Comment by discordier** [comment]
On another note, wouldn't it be better to inject the values of the breadcrumb in [BaseView::createEditMask](https://github.com/contao-community-alliance/dc-general/blob/b719767b9cb7ba64ed65d85bbd48331ae3e1174d/src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/BaseView.php#L776) with a setter function or as constructor argument?
Then we would have the breadcrumb method only located at one class.

```php
        $editMask = new EditMask($this, $model, $originalModel, $preFunction, $postFunction);
        $editMask->setBreadcrumb($this->breadcrumb());
        return $editMask->execute();
```

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php (hunk 5 above)
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: Ok, found it, is already in the original method in base view... Will change it.

**Comment by cboelter** [comment]
@discordier thanks for your feedback. I will have a look on it later this day.

**Comment by cboelter** [comment]
As proposed by @discordier I removed the breadcrumb logic from the edit mask and passed the rendered breadcrumb to it via constructor.

**Comment by cboelter** [comment]
@discordier is everything fine now?

**Pull request merged**